File: Basic/List/List.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func removeDuplicateValues(list_int []int) []int {
	keys := make(map[int]bool)
	list := []int{}

	for _, item := range list_int {
		if _, value := keys[item]; !value {
			keys[item] = true
			list = append(list, item)
		}
	}

	return list
}

func Contains(list_str []string, x string) bool {
	for _, n := range list_str {
		if x == n {
			return true
		}
	}
	return false
}

func Find(list_str []string, x string) int {
	for i, n := range list_str {
		if x == n {
			return i
		}
	}
	return -1
}

func removeValueinList(a []string, index int) []string {
	return append(a[:index], a[index+1:]...)
}

func ContainsKey(m map[int]string, key int) bool {
	if _, ok := m[key]; ok {
		if ok {
			return true
		}
	}

	return false
}

func ContainsValue(m map[int]string, v string) bool {
	for _, x := range m {
		if x == v {
			return true
		}
	}

	return false
}

func main() {

	// make string "Hello World" and string "11223344"
	str1 := "Hello World"
	int1 := "1122334455"

	fmt.Println("Str:", str1, "Int:", int1)
	// turn string into list

	chars := []rune(str1)
	list_str := []string{}

	for i := 0; i < len(chars); i++ {
		list_str = append(list_str, string(chars[i]))
	}

	fmt.Println(list_str)

	// turn list int into list
	chars = []rune(int1)
	arr_int := []string{}

	for i := 0; i < len(chars); i++ {
		arr_int = append(arr_int, string(chars[i]))
	}
	fmt.Println(arr_int)

	// turn list  string into int
	list_int := []int{}

	for _, item := range arr_int {
		j, err := strconv.Atoi(item)
		if err != nil {
			panic(err)
		}

		list_int = append(list_int, j)
	}
	// remove duplicate
	list_int = removeDuplicateValues(list_int)
	fmt.Println(list_int)

	// make another list
	list_int1 := []int{6, 7, 8, 9, 10}

	// merge 2 list with same type
	list_int = append(list_int, list_int1...)

	fmt.Println(list_int)

	//print list into string
	fmt.Println(strings.Join(list_str, ""))

	// check if 'space' inside list
	fmt.Println(Contains(list_str, " "))

	// check index space
	ind_space := Find(list_str, " ")
	fmt.Println(ind_space)

	// remove space in list
	list_str = removeValueinList(list_str, ind_space)
	fmt.Println(list_str)

	// swap list
	list_str[4], list_str[5] = list_str[5], list_str[4]
	fmt.Println(list_str)

	// cut list into 2 left and right
	mid := int(math.Floor(float64(len(list_str) / 2)))

	left, right := []string{}, []string{}

	left, right = append(left, list_str[0:mid]...), append(right, list_str[mid:]...)
	fmt.Println(left, right)

	// make hash map
	m := map[int]string{1: "first", 2: "second", 3: "third"}
	fmt.Println("Map:", m)

	// check if n in keys
	fmt.Println(ContainsKey(m, 1))

	//print hash[index]
	fmt.Println(m[1])
	// check if n in values
	fmt.Println(ContainsValue(m, "first"))

	// itterate

	for key, val := range m {
		fmt.Println("Key:", key, "value:", val)
	}
}

File: Basic/List/List.js
// make string "Hello World" and string "1122334455"
let str1 = "Hello World";
let int1 = "1122334455";

console.log("String :"+str1 + " Integer:"+int1);

// turn string into list
let list_str = str1.split("");
let list_int = int1.split("");

console.log("String :"+list_str + " Integer:"+list_int);

// turn list string into list int
list_int = list_int.map(Number)
console.log(list_int);

// remove duplicate
list_int = [...new Set(list_int)]
console.log(list_int);

// turn list into string
console.log(list_str.join(""))

// check if 'space' inside list
console.log(list_str.includes(" "));

// check index space
let ind_space = list_str.indexOf(" ")
console.log(ind_space)

// remove space in list
list_str.splice(ind_space,1);
console.log(list_str);

// swap list
[list_str[4], list_str[5]] = [list_str[5], list_str[4]];
console.log(list_str);

// make another list
let list_int1 = [6,7,8,9,10]


// merge 2 list with same type
list_int = list_int.concat(list_int1);
console.log(list_int);


// cut list into 2 left and right
mid = list_str.length/2
left = list_str.splice(0, mid);
right = list_str.splice(-mid);
console.log(left);
console.log(right);

// make hash map
let hash_map = new Map([
    [1, "first"],
    [2, "second"],
    [3, "third"],
])

// check if n in keys
console.log(hash_map.has(1));

// get second value
console.log(hash_map.get(2));

// check if n in values
console.log(Array.from(hash_map.values()).includes("third"));

//itterate hashmap
for(let [key, val] of hash_map){
    console.log(key+ " = "+ val);
}
